crates/assists/src/handlers/inline_function.rs

```diff
@@ -0,0 +1,96 @@
+use ast::make;
+use hir::{HasSource, HasVisibility, ModuleDef, PathResolution};
+use syntax::{
+    ast::{self, edit::AstNodeEdit, ArgListOwner},
+    AstNode,
+};
+
+use crate::{
+    assist_context::{AssistContext, Assists},
+    AssistId, AssistKind,
+};
+
+// Assist: inline_function
+//
+// Inlines a function body.
+//
+// ```
+// fn add(a: u32, b: u32) -> u32 { a + b }
+// fn main() {
+//     let x = add<|>(1, 2);
+// }
+// ```
+// ->
+// ```
+// fn add(a: u32, b: u32) -> u32 { a + b }
+// fn main() {
+//     let x = {
+//         let a = 1;
+//         let b = 2;
+//         a + b
+//     };
+// }
+// ```
+pub(crate) fn inline_function(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let path_expr: ast::PathExpr = ctx.find_node_at_offset()?;
+    let call = path_expr.syntax().parent().and_then(ast::CallExpr::cast)?;
+    let path = path_expr.path()?;
+
+    let function = match ctx.sema.resolve_path(&path)? {
+        PathResolution::Def(ModuleDef::Function(f)) => f,
+        _ => return None,
+    };
+
+    let current_scope = ctx.sema.scope(call.syntax());
+    let current_module = current_scope.module()?;
+
+    if !function.is_visible_from(ctx.db(), current_module) {
+        // The function isn't accessible from here so we can't inline it
+        return None;
+    }
+
+    let function_source = function.source(ctx.db());
+    let body = function_source.value.body()?.expr()?;
```

Comment:
Looks like this'll strip most of the code away? I think the easiest would be to implement `ast::BlockExpr::preprend_stmts` in `edit.rs`

Reply:
Thanks, I didn't know that `expr()` will only return the last expression in the function and was thinking it'd return the entire body as a `BlockExpr`.

I've added a test which includes a multi-statement function and fixed the code.

Reply:
hm, we should rename `BlockExpr::expr`...

Reply:
https://github.com/rust-analyzer/rust-analyzer/pull/7168